test/videoframeprocessor/DefaultVideoTransformDevice.test.ts

```diff
@@ -0,0 +1,444 @@
+        destroy(): Promise<void> {
+          return;
+        }
+        process(_buffers: VideoFrameBuffer[]): Promise<VideoFrameBuffer[]> {
+          throw new Error('Method not implemented.');
+        }
+      }
+
+      const obs = new MockObserver();
+      const failToStartCallback = new Promise(resolve => {
+        obs.processingDidFailToStart.callsFake(resolve);
+      });
+
+      const processor = new WrongProcessor();
+      const device = new DefaultVideoTransformDevice(logger, 'test', [processor]);
+      device.addObserver(obs);
+      await device.transformStream(mockVideoStream);
+
+      await failToStartCallback;
+      await device.stop();
+    });
+
+    it('processingDidFailToStart with multiple observers', async () => {
+      class WrongProcessor implements VideoFrameProcessor {
+        destroy(): Promise<void> {
+          return;
+        }
+        process(_buffers: VideoFrameBuffer[]): Promise<VideoFrameBuffer[]> {
+          throw new Error('Method not implemented.');
+        }
+      }
+
+      class Observer implements DefaultVideoTransformDeviceObserver {
+        processingDidFailToStart = stub();
+      }
+
+      class Observer2 implements DefaultVideoTransformDeviceObserver {
+        processingDidStart = stub();
+      }
+
+      const obs = new Observer();
+      const startCallback = new Promise(resolve => {
+        obs.processingDidFailToStart.callsFake(resolve);
+      });
+
+      const processor = new WrongProcessor();
+      const device = new DefaultVideoTransformDevice(logger, 'test', [processor]);
+      device.addObserver(obs);
+      const obs2 = new Observer2();
+      device.addObserver(obs2);
+      await device.transformStream(mockVideoStream);
+
+      await startCallback;
+      await device.stop();
+      expect(obs2.processingDidStart.called).to.equal(false);
+    });
+
+    it('processingDidFailToStart', async () => {
+      class WrongProcessor implements VideoFrameProcessor {
+        destroy(): Promise<void> {
+          return;
+        }
+        process(_buffers: VideoFrameBuffer[]): Promise<VideoFrameBuffer[]> {
+          throw new Error('Method not implemented.');
+        }
+      }
+      const obs = new MockObserver();
+      const failToStartCallback = new Promise(resolve => {
+        obs.processingDidFailToStart.callsFake(resolve);
+      });
+      const processor = new WrongProcessor();
+      const device = new DefaultVideoTransformDevice(logger, 'test', [processor]);
+      device.addObserver(obs);
+      await device.transformStream(mockVideoStream);
+      await failToStartCallback;
+      await device.stop();
+
+      device.removeObserver(obs);
+    });
+
+    it('processingLatencyTooHigh', async () => {
+      class Observer2 implements DefaultVideoTransformDeviceObserver {}
+      const obs = new MockObserver();
+
+      const latencyCallback = new Promise(resolve => {
+        obs.processingLatencyTooHigh.callsFake(resolve);
+      });
```

Comment:
See my comment earlier for a helper function to remove a bunch of duplicated code around each of these.